bastion.cfn.yml

```diff
@@ -16,68 +16,290 @@ Parameters:
+                  echo "AuthenticationMethods publickey,keyboard-interactive" >> /etc/ssh/sshd_config
+                  service sshd restart
+                fi
+                rm -f /usr/local/sbin/configure_mfa.sh
+
+          commands:
+            configure-mfa:
+              command: "/usr/local/sbin/configure_mfa.sh"
+
+          services:
+              sysvinit:
+                cfn-hup:
+                  enabled: true
+                  ensureRunning: true
+                  files:
+                    - /etc/cfn/cfn-hup.conf
+                    - /etc/cfn/hooks.d/cfn-auto-reloader.conf
+                awslogs:
+                  enabled: true
+                  ensureRunning: true
+                  files:
+                    - /etc/awslogs/awslogs.conf
+
     Properties:
       InstanceType: t2.micro
       KeyName: !Ref KeyName
+      NetworkInterfaces:
+        -
+          NetworkInterfaceId: !Ref BastionNetworkInterface
+          DeviceIndex: 0
+
       ImageId: !FindInMap [AMIMap, !Ref "AWS::Region", AMI]
-      SubnetId: !ImportValue
-        "Fn::Sub": "${NetworkStackName}-PublicSubnet1ID"
-      SecurityGroupIds:
-      - !ImportValue
-        "Fn::Sub": "${NetworkStackName}-BastionGroupID"
+      UserData:
+        Fn::Base64: !Sub |
+          #!/bin/bash -xe
+          yum update -y
+          /opt/aws/bin/cfn-init -v -s ${AWS::StackId} --resource BastionHost --region ${AWS::Region}
+          /opt/aws/bin/cfn-signal -e $? --stack ${AWS::StackId} --resource BastionHost --region ${AWS::Region}
+
+      IamInstanceProfile: !Ref BastionInstanceProfile
       Tags:
         -
           Key: Name
           Value: startup-kit-bastion
+    DependsOn: BastionSshPort
+    CreationPolicy:
+      ResourceSignal:
+        Count: 1
+        Timeout: PT5M
 
   BastionEIP:
     Type: AWS::EC2::EIP
     Properties:
-      InstanceId: !Ref BastionHost
       Domain: vpc
 
+  BastionSshPort:
+    Type: AWS::EC2::EIPAssociation
+    Properties:
+      AllocationId: !GetAtt BastionEIP.AllocationId
+      NetworkInterfaceId: !Ref BastionNetworkInterface
+    DependsOn:
+      - BastionEIP
+      - BastionNetworkInterface
+
+  BastionNetworkInterface:
+    Type: AWS::EC2::NetworkInterface
+    Properties:
+      SubnetId:
+        Fn::ImportValue: !Sub "${NetworkStackName}-PublicSubnet1ID"
+      GroupSet:
+        - Fn::ImportValue: !Sub "${NetworkStackName}-BastionGroupID"
+      SourceDestCheck: true
+      Tags:
+        -
+          Key: Name
+          Value: startup-kit-bastion
+
+  BastionSecureLogGroup:
+    Type: AWS::Logs::LogGroup
+    Properties:
+      RetentionInDays: !Ref LogRetentionInDays
+
+  BastionSecureLogGroupStream:
+    Type: AWS::Logs::LogStream
+    Properties:
+      LogGroupName: !Ref BastionSecureLogGroup
+      LogStreamName: log
+
+
+# InvalidUser and InvalidUserAlarm:
+# When a user tries to SSH with invalid username the next line is logged in the SSH log file:
+  InvalidUser:
```

Comment:
I assume our guidance is to never use 0.0.0.0/0 -> TCP:22 for bastion hosts, but it's worth noting that if a user does that, these alarms will trigger constantly. Not necessarily an issue (and probably A Good Thing), but perhaps worth a note in the README.